File: adt2unit/i386/is_4op_chan.pas
// This file is part of Adlib Tracker II (AT2).
//
// SPDX-FileType: SOURCE
// SPDX-FileCopyrightText: 2014-2024 The Adlib Tracker 2 Authors
// SPDX-License-Identifier: GPL-3.0-or-later

function is_4op_chan(chan: Byte): Boolean;

var
  result: Boolean;

begin
  asm
        mov     al,byte ptr [songdata.flag_4op]
        mov     ah,chan
        test    al,1
        jz      @@1
        cmp     ah,1
        jb      @@1
        cmp     ah,2
        ja      @@1
        mov     result,TRUE
        jmp     @@7
@@1:    test    al,2
        jz      @@2
        cmp     ah,3
        jb      @@2
        cmp     ah,4
        ja      @@2
        mov     result,TRUE
        jmp     @@7
@@2:    test    al,4
        jz      @@3
        cmp     ah,5
        jb      @@3
        cmp     ah,6
        ja      @@3
        mov     result,TRUE
        jmp     @@7
@@3:    test    al,8
        jz      @@4
        cmp     ah,10
        jb      @@4
        cmp     ah,11
        ja      @@4
        mov     result,TRUE
        jmp     @@7
@@4:    test    al,10h
        jz      @@5
        cmp     ah,12
        jb      @@5
        cmp     ah,13
        ja      @@5
        mov     result,TRUE
        jmp     @@7
@@5:    test    al,20h
        jz      @@6
        cmp     ah,14
        jb      @@6
        cmp     ah,15
        ja      @@6
        mov     result,TRUE
        jmp     @@7
@@6:    mov     result,FALSE
@@7:
  end;
  is_4op_chan := result;
end;

File: adt2unit/is_4op_chan.c
// This file is part of Adlib Tracker II (AT2).
//
// SPDX-FileType: SOURCE
// SPDX-FileCopyrightText: 2014-2024 The Adlib Tracker 2 Authors
// SPDX-License-Identifier: GPL-3.0-or-later

// chan: 1..20
bool is_4op_chan (uint8_t chan) {
  chan--;

  return (        (songdata.flag_4op & 0x3F)
          && (   ((songdata.flag_4op & (1 << 0)) && ((chan ==  0) || (chan ==  1)))
              || ((songdata.flag_4op & (1 << 1)) && ((chan ==  2) || (chan ==  3)))
              || ((songdata.flag_4op & (1 << 2)) && ((chan ==  4) || (chan ==  5)))
              || ((songdata.flag_4op & (1 << 3)) && ((chan ==  9) || (chan == 10)))
              || ((songdata.flag_4op & (1 << 4)) && ((chan == 11) || (chan == 12)))
              || ((songdata.flag_4op & (1 << 5)) && ((chan == 13) || (chan == 14)))));
}
